### ISSUE TITLE
preserve colors in diagnostics

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust-playground#1117]
rustc uses a lot of colors and highlighting in its errors. sometimes the output is hard to read when those are removed. consider for example this diagnostic, where it's hard to parse that the `-` is meant to be a highlight: ![Image](https://github.com/user-attachments/assets/9064769c-2147-40b8-821b-91b2f3e3aaa7)

it would be nice if playground preserved the colors.

repro [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=fn+main%28%29+%7B%0A++std%3A%3Aio.lines%28%29%3B%0A%7D)

**Comment by estebank** [comment]
It might be enough to use `anstyle-svg` for this, but that would require additional work for the existing links. A likely better option would be to have the json output emit the colored ANSI output and use some other crate to turn the codes into HTML that we can style.

**Comment by shepmaster** [comment]
> rustc uses a lot of colors and highlighting in its errors. sometimes the output is hard to read when those are removed

Should that be a general concern for people who don’t have colors for whatever given reason?

**Comment by estebank** [comment]
@shepmaster I make an effort to make sure that all output is readable as text only, but undoubtedly color makes the output easier to scan and reduces the likelihood of misreading things. In the screenshot, someone very knowledgeable mistook the `-` as a connector between the `^^` and `help...`. With colors, the `help` appears blue, like the `-`, making it easier to differentiate from the `^^`. That being said, this suggestion in particular is also being changed to be verbose style so that we no longer have this confusion. I want us to move away from inline suggestions entirely (but that requires me building a PR while `--bless`ing `stderr` files for every platform, which is a bit of a pain), but that will still leave some corner cases where non-suggestion short labels can be misread slightly.

**Comment by jyn514** [comment]
> Should that be a general concern for people who don’t have colors for whatever given reason?

note that this particular case is being fixed upstream in rustc: https://github.com/rust-lang/rust/pull/132780

but it would be nice to fix it in the general case as well; it's never going to be possible to communicate *as much* information as clearly with text alone.

**Comment by estebank** [comment]
I tried some time back making all suggestions have the verbose diff output in https://github.com/rust-lang/rust/pull/127282. Landing that change or one like it will be a huge pain in the neck.

**Comment by shepmaster** [comment]
I can't find an issue where I've discussed this before, but the big thing that I want from a feature like this is to not directly trust the ANSI codes. `cargo` and `rustc` will not have an idea about the color scheme that the Playground's user has selected, so whatever colors they output need to be treated more abstractly so the frontend can use them in a structured and themed way.

Ideally, there'd be a way to get output from running the build in that structured form, before it was highlighted, but I don't know of anything like that.
